test/Rings/MPolyAnyMap/flattenings.jl

```diff
@@ -117,3 +115,198 @@ end
+
+  I = ideal(S, [u*(x^2 + y^2)])
+  I_flat = flat(I)
+  @test u in I
+  @test !(v in I)
+end
+
+@testset "flattenings of modules" begin
+  R, (x, y) = polynomial_ring(QQ, [:x, :y], cached=false)
+  S, (u, v) = grade(polynomial_ring(R, [:u, :v], cached=false)[1])
+
+  flat = Oscar.flatten(S)
+
+  S_flat = codomain(flat)
+
+  F = graded_free_module(S, [-2, 1])
+  F_flat, iso, iso_inv = Oscar.flatten(F)
+
+  M, _ = sub(F, [F[1]])
+  M_flat, iso_M, iso_M_inv = Oscar.flatten(M)
+  @test is_graded(M_flat)
+  @test ambient_free_module(M_flat) === F_flat
+  @test F[1] in M
+  @test coordinates(u*F[1], M)[1] == u
+
+  @test is_graded(F_flat)
+
+  @test iso_inv(iso(F[1])) == F[1]
+  @test F_flat === Oscar.flatten(F)[1]
+
+  A, _ = quo(R, [x-y])
+
+  S, (u, v) = grade(A[:u, :v][1])
+
+  flat = Oscar.flatten(S)
+
+  S_flat = codomain(flat)
+
+  F = graded_free_module(S, [-2, 1])
+  F_flat, iso, iso_inv = Oscar.flatten(F)
+
+  M, _ = sub(F, [F[1]])
+  M_flat, iso_M, iso_M_inv = Oscar.flatten(M)
+  @test is_graded(M_flat)
+  @test ambient_free_module(M_flat) === F_flat
+  @test F[1] in M
+  @test coordinates(u*F[1], M)[1] == u
+
+  @test iso_inv(iso(F[1])) == F[1]
+  @test F_flat === Oscar.flatten(F)[1]
+  @test is_graded(F_flat)
+
+  U = powers_of_element(y)
+  L, _ = localization(R, U)
+  S, (u, v) = grade(L[:u, :v][1])
+
+  flat = Oscar.flatten(S)
+  S_flat = codomain(flat)
+  U, V, X, Y = gens(S_flat)
+
+  @test degree(U) == degree(u)
+  @test iszero(degree(X))
+
+  F = graded_free_module(S, [-2, 1])
+  F_flat, iso, iso_inv = Oscar.flatten(F)
+
+  M, _ = sub(F, [F[1]])
+  M_flat, iso_M, iso_M_inv = Oscar.flatten(M)
+  @test is_graded(M_flat)
+  @test ambient_free_module(M_flat) === F_flat
+  @test F[1] in M
+  @test coordinates(u*F[1], M)[1] == u
+
+  @test iso_inv(iso(F[1])) == F[1]
+  @test F_flat === Oscar.flatten(F)[1]
+  @test is_graded(F_flat)
+
+  U = powers_of_element(y)
+  W, _ = localization(A, U)
+  S, (u, v) = grade(W[:u, :v][1])
+
+  flat = Oscar.flatten(S)
+  S_flat = codomain(flat)
+  U, V, X, Y = gens(S_flat)
+
+  @test degree(U) == degree(u)
+  @test iszero(degree(X))
+
+  F = graded_free_module(S, [-2, 1])
+  F_flat, iso, iso_inv = Oscar.flatten(F)
+
+  M, _ = sub(F, [F[1]])
+  M_flat, iso_M, iso_M_inv = Oscar.flatten(M)
+  @test is_graded(M_flat)
+  @test ambient_free_module(M_flat) === F_flat
+  @test F[1] in M
+  @test coordinates(u*F[1], M)[1] == u
+
+  @test iso_inv(iso(F[1])) == F[1]
+  @test F_flat === Oscar.flatten(F)[1]
+  @test is_graded(F_flat)
+end
+
+@testset "free (graded) resolutions" begin
+  R, (x, y) = polynomial_ring(QQ, [:x, :y], cached=false)
+  S, (u, v) = grade(polynomial_ring(R, [:u, :v], cached=false)[1])
+
+  flat = Oscar.flatten(S)
+
+  S_flat = codomain(flat)
+
+  F = graded_free_module(S, [-2, 1])
+  F_flat, iso, iso_inv = Oscar.flatten(F)
+
+  M, _ = quo(F, [u^5*F[1], v*F[1]])
+  M_flat, iso, iso_inv = Oscar.flatten(M)
+  res = free_resolution(M)
```

Comment:
@jankoboehm : Could you maybe have a look at this and tell me whether I implemented this custom free resolution in accordance with how you guys designed this? Maybe just run the code of this test in the REPL once, or something?